### PR TITLE
feat: Establish Context for filtering

### PR DESCRIPTION
**Pull request by arnest00**
### Description
- `context/FilterContext.jsx` exports:
  - `useFilterContext` custom hook
  - `FilterProvider` custom Provider
- `FilterProvider` wraps around Overview page
- TradingCardGrid consumes context through `useFilterContext`
- `getUniqueCurrentProjects` and `getUniqueRoles` util functions
  - Test behavior of both functions

### Spec
See Story: [228](https://sparkbox.atlassian.net/browse/FSA22V1-228)

### Validation
* [ ] This PR has code changes, and our linters still pass.
* [ ] This PR affects production code, so it was browser tested.
* [ ] This PR has new code, so new tests were added or updated, and they pass.
* [ ] Changes are browser tested. This includes functionality, accessibility, responsiveness, and design.

#### To Validate
1. Make sure all PR Checks have passed.
2. Pull down `feat--use-context` branch.
3. Open deploy preview and verify that Merry, the only team member without availability, is not included in the Overview page.
4. Verify all tests pass.

### CODE REVIEW

File: components/TradingCardGrid.jsx [hunk 1]
```diff
@@ -1,10 +1,22 @@
...
+  const filteredTeamMembersByRole = filteredTeamMembersByAvailability
+    .filter((member) => roles.includes(member.jobTitle));
+
+  const filteredTeamMembers = filteredTeamMembersByRole
+    .filter((member) => project === 'all' || member.project === project);
+
   return (
     <div className="cmp-trading-card-grid">
-      {teamMembers.map((member) => (
+      {filteredTeamMembers.map((member) => (
```
Comment (arnest00) [inline components/TradingCardGrid.jsx:19]: `filteredTeamMembers`, using the default state I've included in `FilterContext.jsx`, includes available team members from all roles across all projects. (Sorry, Merry.)

File: context/FilterContext.jsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  const [roles, setRoles] = useState([
+    'Developer',
+    'Director of Business Development',
+    'Frontend Designer',
+    'Project Manager',
+    'Technical Director',
+    'UI Designer',
+    'User Experience Strategist',
+    'Projects Director',
+  ]);
```
Comment (arnest00) [inline context/FilterContext.jsx:21]: All of this will be refactored as we figure out what the controls are going to give us, but I wanted to have something in here to show that the custom hook was working.

File: util/getUniqueRoles.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+  for (let i = 0; i < teamMemberIdArray.length; i += 1) {
+    const thisRole = teamMembers[teamMemberIdArray[i]].jobTitle;
+
+    if (!uniqueRoles.includes(thisRole)) {
+      uniqueRoles.push(thisRole);
+    }
+  }
+
+  return uniqueRoles;
```
Comment (arnest00) [inline util/getUniqueRoles.js:13]: Currently no sorting being done with roles.

File: components/TradingCardGrid.jsx [hunk 4]
```diff
@@ -1,10 +1,22 @@
...
+  const { project, roles, availability } = useFilterContext();
+
+  const filteredTeamMembersByAvailability = teamMembers
+    .filter((member) => ((member.weeklyCapacity - member.forecastedHours) > 0) === availability);
+
+  const filteredTeamMembersByRole = filteredTeamMembersByAvailability
+    .filter((member) => roles.includes(member.jobTitle));
+
+  const filteredTeamMembers = filteredTeamMembersByRole
+    .filter((member) => project === 'all' || member.project === project);
```
Comment (dustin-jw) [inline components/TradingCardGrid.jsx:15]: We ought to be able to combine these into a single function, returning false any time we don't match a condition, then if we make it to the end, we return true:

```suggestion
  const filteredTeamMembers = teamMembers.filter((member) => {
    if (member.weeklyCapacity - member.forcastedHours <= 0) return false;
    
    if (!roles.includes(member.jobTitle)) return false;
    
    if (project !== 'all' && member.project !== project) return false;
    
    return true;
  });
```

**Review by dustin-jw — CHANGES_REQUESTED** [review]
Not a full review, but something I noticed that could be refactored fairly easily.

File: components/TradingCardGrid.jsx (hunk 4 above)
Comment (arnest00) [reply components/TradingCardGrid.jsx:15]: Would the following still make sense, assuming we keep availability as a boolean and that we are still toggling between available and non-available team members?

```js
  const filteredTeamMembers = teamMembers.filter((member) => {
    if ((member.weeklyCapacity - member.forcastedHours <= 0) === availability) return false;

    if (!roles.includes(member.jobTitle)) return false;

    if (project !== 'all' && member.project !== project) return false;

    return true;
  });
```

Availability is true: positve numbers evaluate to false, availability is true, member included.
Availability is true: 0 evaluates to true, availability is true, member not included.
Availability is false: positive numbers evaluate to false, availability is false, member not included.
Availability is false: 0 evaluates to true, availability is false, member included.

Comment (dustin-jw) [reply components/TradingCardGrid.jsx:15]: That's a clever way to do it, but I pretty much always favor clarity over cleverness. I'm also thinking that using a boolean for `availability` is somewhat limiting. I can see a future where we would want a 'both' option or a 'half-time' option, which would be easier to add if we started with using strings. What if we did something like this?

```js
const hasHoursAvailable = member.weeklyCapacity - member.forecastedHours > 0;
if (!hasHoursAvailable && availability === 'available') return false;

if (hasHoursAvailable && availability === 'unavailable') return false;
```

File: util/getUniqueCurrentProjects.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+  for (let i = 0; i < teamMemberIdArray.length; i += 1) {
+    for (let j = 0; j < teamMembers[teamMemberIdArray[i]].currentProjects.length; j += 1) {
+      const thisProject = teamMembers[teamMemberIdArray[i]].currentProjects[j].project;
+
+      if (!uniqueCurrentProjects.includes(thisProject)) {
+        uniqueCurrentProjects.push(thisProject);
+      }
+    }
+  }
```
Comment (dustin-jw) [inline util/getUniqueCurrentProjects.js:13]: I think we can simplify this a bit by using [flat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) and [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#remove_duplicate_elements_from_the_array).

```js
const projectsList = Object.keys(teamMembers).map((key) => teamMembers[key].currentProjects);

const uniqueProjects = [...new Set(projectsList.flat())];
```

This should remove duplicates and _may_ sort it automatically, but definitely test it out before removing the sorting methods.

File: util/getUniqueCurrentProjects.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+  // Sorts uniqueCurrentProjects alphabetically
+  const sortedUniqueCurrentProjects = uniqueCurrentProjects
+    .sort((a, b) => {
+      if (a < b) return -1;
+      if (a > b) return 1;
+      return 0;
+    });
```
Comment (dustin-jw) [inline util/getUniqueCurrentProjects.js:21]: If we're only sorting a list of strings alphabetically, we don't need to implement the sorting logic – `.sort()` will work as-is 😃 

Normally I save the `a < b` type of stuff if I'm sorting properties of objects or things like that.

File: util/getUniqueRoles.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  const teamMemberIdArray = Object.keys(teamMembers);
+  const uniqueRoles = [];
+
+  for (let i = 0; i < teamMemberIdArray.length; i += 1) {
+    const thisRole = teamMembers[teamMemberIdArray[i]].jobTitle;
+
+    if (!uniqueRoles.includes(thisRole)) {
+      uniqueRoles.push(thisRole);
+    }
+  }
```
Comment (dustin-jw) [inline util/getUniqueRoles.js:11]: We can do some similar stuff with `Set` and mapping to simplify this as well:

```js
const roles = Object.keys(teamMembers).map((key) => teamMembers[key].jobTitle);

const uniqueRoles = [...new Set(roles)];
```

I can't overstate just how magical `Set` is.

**Review by dustin-jw — CHANGES_REQUESTED** [review]
I couldn't resist wanting to refactor out those `for` loops 😄  Let me know if you have any questions about those suggestions, there are a lot of cool array methods but they aren't always intuitive.

File: context/FilterContext.jsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import {
+  createContext, useContext, useMemo, useState,
+} from 'react';
+
+const FilterContext = createContext();
+
+export const useFilterContext = () => useContext(FilterContext);
+
+export function FilterProvider({ children }) {
+  const [availability, setAvailability] = useState(true);
```
[the remark this answers is not in the document]
Comment (arnest00) [reply context/FilterContext.jsx:10]: I'll export that object too, so that the future Availability component can access it.

File: util/getUniqueCurrentProjects.js (hunk 6 above)
Comment (arnest00) [reply util/getUniqueCurrentProjects.js:21]: TIL: default `sort` behavior

File: util/getUniqueCurrentProjects.js (hunk 5 above)
Comment (arnest00) [reply util/getUniqueCurrentProjects.js:13]: TIL: `Set`
TIL: `.flat()`

File: components/TradingCardGrid.jsx [hunk 9]
```diff
@@ -1,10 +1,26 @@
...
 import TradingCard from './TradingCard';
 
 export default function TradingCardGrid({ teamMembers }) {
+  const { project, roles, availability } = useFilterContext();
+
+  const filteredTeamMembers = teamMembers.filter((member) => {
+    const hasHoursAvailable = member.weeklyCapacity - member.forecastedHours > 0;
+
+    if (!hasHoursAvailable && availability === 'available') return false;
+    if (hasHoursAvailable && availability === 'unavailable') return false;
```
Comment (dustin-jw) [inline components/TradingCardGrid.jsx:12]: I would import `availabilityOptions` here, so we can use it in our comparisons:

```js
    if (!hasHoursAvailable && availability === availabilityOptions.AVAILABLE) return false;
    if (hasHoursAvailable && availability === availabilityOptions.UNAVAILABLE) return false;
```

File: context/FilterContext.jsx [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  const [roles, setRoles] = useState([
+    'Developer',
+    'Director of Business Development',
+    'Frontend Designer',
+    'Project Manager',
+    'Technical Director',
+    'UI Designer',
+    'User Experience Strategist',
+    'Projects Director',
+  ]);
```
Comment (dustin-jw) [inline context/FilterContext.jsx:26]: This should work in such a way that if `roles` is an empty array (i.e. no checkboxes checked), then we should see everybody. Let's update this to be `useState([])`, then I'll leave a comment in the logic on how we could achieve our intended behavior.

File: components/TradingCardGrid.jsx [hunk 11]
```diff
@@ -1,10 +1,26 @@
...
+
+    if (!roles.includes(member.jobTitle)) return false;
```
Comment (dustin-jw) [inline components/TradingCardGrid.jsx:14]: Since "no roles selected" should behave the same way as if all roles are selected, we should update this accordingly.

```js
    if (roles.length && !roles.includes(member.jobTitle)) return false;
```

File: util/getUniqueCurrentProjects.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+export default function getUniqueCurrentProjects(teamMembers) {
+  const projectsList = Object.keys(teamMembers)
+    .map((key) => teamMembers[key].currentProjects)
+    .flat();
+
+  const projectNamesList = projectsList.map((project) => project.project);
+
+  const sortedUniqueProjects = [...new Set(projectNamesList)].sort();
```
Comment (dustin-jw) [inline util/getUniqueCurrentProjects.js:8]: Very nice! We could split this up into two steps to separate the getting-unique-values part from the sorting part, but I don't feel super strongly either way.

**Review by dustin-jw — CHANGES_REQUESTED** [review]
This is super close, basically just one functional change and some minor code style things to work through, and it should be good to go! Preemptively getting @theogainey to put eyes on this before I approve.

File: util/getUniqueCurrentProjects.js (hunk 12 above)
Comment (arnest00) [reply util/getUniqueCurrentProjects.js:8]: @theogainey decide this for us

Comment (theogainey) [reply util/getUniqueCurrentProjects.js:8]: I think it would be best to spilt it up. Let's say someone decides it would be a good idea to sort the projects a different way, having a clear sort step would make it easier for that someone to implement that.

Comment (arnest00) [reply util/getUniqueCurrentProjects.js]: 🏆

**Review by dustin-jw — DISMISSED** [review]

File: util/getUniqueCurrentProjects.test.js [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+      ],
+    },
+  };
+
+  it('returns unique current projects from all team members as an array in alphabetical order', () => {
+    const currentProjects = getUniqueCurrentProjects(MOCK_TEAM_MEMBERS_DATA);
+
+    expect(currentProjects).toEqual(['Dol Guldur', 'Lonely Mountain Residency', 'Quest of Erebor']);
+  });
+});
```
Comment (theogainey) [inline util/getUniqueCurrentProjects.test.js:48]: I think it would be smart to add some tests for some edge cases. Some thoughts, what if a currentProjects is empty `currentProjects : []`, what if the a project doesn't have a name just a client. What if a there are some null or undefined values floating around in the data. We want to make sure this code still behaves as expected in these cases.

File: util/getUniqueRoles.test.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+      jobTitle: 'Project Manager',
+    },
+  };
+
+  it('returns unique roles from all team members as an array', () => {
+    const roles = getUniqueRoles(MOCK_TEAM_MEMBERS_DATA);
+
+    expect(roles).toEqual(['Developer', 'UI Designer', 'Technical Director', 'Project Manager']);
+  });
+});
```
Comment (theogainey) [inline util/getUniqueRoles.test.js:30]: Same goes for here. Let's add some tests for edge cases. Our data isn't always going to place nice with use so that make sure we know this code can handle.

**Review by theogainey — CHANGES_REQUESTED** [review]
Good work, `context` is hard and this code looks like it's going to play nice with what we need it for. But seeing as we are going to be using this `context` for a few different things I think it would be smart to add a more robust set of tests to this code. That way if we accidentally break things down the line it will be much easier to find out where we went wrong.

File: util/getUniqueCurrentProjects.test.js (hunk 13 above)
Comment (dustin-jw) [reply util/getUniqueCurrentProjects.test.js:48]: 🏆 Good looking out!

File: util/getUniqueCurrentProjects.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
+export default function getUniqueCurrentProjects(teamMembers) {
```
Comment (arnest00) [inline util/getUniqueCurrentProjects.js:1]: Making some assumptions for app behavior here:
- If there is no client and no project specified, we don't want to include that in the select.
- If there is a client but no project, we will include that as 'Unnamed `client` Project'.

File: util/getUniqueRoles.js [hunk 16]
```diff
@@ -0,0 +1,12 @@
+export default function getUniqueRoles(teamMembers) {
```
Comment (arnest00) [inline util/getUniqueRoles.js:1]: Ditto here:
- If a role evaluates to a truthy value, we will include that in the list of roles.
- If a role is stored as an unexpected data type, we will not include that in the list of roles.

**Review by theogainey — CHANGES_REQUESTED** [review]
Good work thinking about what assumptions we can make about the data and the intentions on how we want to handle those assumptions. Now let's communicate those assumptions in our tests. Instead of just one test that uses a combination of good and bad data lets break it down into a set of tests. 

1. Nice data nice result `it( returns X unique values )`
2. Nice data nice result `it( returns sorted values )`
3. Edge case 1 `it (  does this for this edge case ) `
4. Repeat 3 until you hit all the edge cases you want
